### ISSUE TITLE
Feature Request - make email unique

### ISSUE THREAD

**Issue by dustymc** [ArctosDB/arctos#7803]
**Is your feature request related to a problem? Please describe.**

Nonunique emails exist, they trip my 'probably violating TOS' scripts by looking like one agent trying to make lots of operator accounts - and some of them probably are just that.

**Describe what you're trying to accomplish** 

Keep things secure, don't violate any of TACC's rules (which are also intended to keep things secure).

**Describe the solution you'd like**

Unique emails.

**Describe alternatives you've considered**

I'm very open to ideas, but I think this hole cannot be dug deeper. 

Attributes have some awareness of type and value, I probably can't just dump these into remarks-or-something.

If I don't get help, I'll do SOMETHING - maybe disem@ them and then dump them somewhere.

PLEASE HELP!

**Additional context**

This has become a near-daily issue, I will do something very soon, clever ideas for dealing with this very appreciated.

I'll open an internal issue for data, it probably shouldn't be shared here.

**Priority**

Critical security-related issue.

**Comment by Jegelewicz** [comment]
Are these duplicates in agent profiles or operator profiles?

**Comment by Jegelewicz** [comment]
That's hilarious? There are definitely old-timers that share(d) an email and then there's that big bunch of OGL agents that all share one. @falco-rk 

I get that ideally every agent would have a unique email address, but in real life that isn't always true. I don't understand how it poses a security risk on the agent profile - it's just data, right? I DO see how two operators with the same email would be problematic. I guess I don't understand the issue properly.

**Comment by falco-rk** [comment]
Can you tell me which OGL agents have the same email? I searched in arctos for our shared email addresses (oceangenome@gmail.com and oglinfo@northeastern.edu) and didn't find anything. I don't have a problem with those only being linked to the OGL genomics collection agent.

**Comment by Jegelewicz** [comment]
@falco-rk they all have this email - imondi[at]coastalmarinebiolabs.org

**Comment by dustymc** [comment]
>real life that isn't always true.

Yea, there are a couple of what might be family accounts, a couple of legit-but-unnecessary "me in some other role" entries, a bunch of duplicates including what look like duplicates with accounts, and what lead to https://github.com/ArctosDB/arctos/issues/7802 (which very likely involved a frustrated new user, but I've got no way to communicate because it was someone else's email).

Schedule a time if this somehow needs more discussion or explanation, if we're going to do that we should include TACC and @lkvoong. (TACC would like operators to have 2FA TACC accounts, that should sufficiently address my security concerns, perhaps we should skip my probably-futile attempts at minimally-intrusive security and just do that?)

@falco-rk data are in the linked internal issue https://github.com/ArctosDB/internal/issues/322 - let us know and someone can email it to you if you don't have access.

**Comment by Jegelewicz** [comment]
Also, some of these are duplicates WITHIN an agent because reasons

PREFERRED_AGENT_NAME | AGENT_ID | ATTRIBUTE_VALUE | address type
-- | -- | --| -- 
Josh R. Pletcher | 21317739 | jrpletch[at]owu.edu | email
Josh R. Pletcher | 21317739 | jrpletch[at]owu.edu | notification email

**Comment by Jegelewicz** [comment]
And some indicate duplicate agents

PREFERRED_AGENT_NAME | AGENT_ID | ATTRIBUTE_VALUE
-- | -- | --
Ellie Bush | 21352392 | bush1[at]ls.byu.edu
Elyssa Bush | 21352394 | bush1[at]ls.byu.edu
Elyssa Bush | 21352393 | bush1[at]ls.byu.edu

**Comment by falco-rk** [comment]
>imondi[at]coastalmarinebiolabs.org

Well, that was an _interesting_ choice we made. That is just one person's email (Ralph Imondi, agent id 21349246). It should be removed from all of the other agents it is associated with.

**Comment by Jegelewicz** [comment]
OK - this one tells me what the problem is

https://arctos.database.museum/agent/21331627

And I am pretty sure I know what causes it - https://handbook.arctosdb.org/how_to/How-to-Create-your-Arctos-Team-Users-and-Operators.html#3-invite-your-user-to-become-an-operator see the Pro Tip

People aren't following through and changing the email addresses....

So maybe this shouldn't be a thing and before any user account can be upgraded to an operator, there must be a unique email associated with the agent? I don't know how hard that would be to check, but also, no two operator accounts should have the same email? (USER accounts can, but not OPERATOR) Possible?

**Comment by Jegelewicz** [comment]
@adhornsby can you please add a unique email account to https://arctos.database.museum/agent/21331627? And update their operator email as well?

**Comment by Jegelewicz** [comment]
This one must go? https://arctos.database.museum/agent/21257829

**Comment by Jegelewicz** [comment]
This is definitely a problem

![image](https://github.com/ArctosDB/arctos/assets/5725767/576ab56b-449b-4bc2-961e-3e5984dd9d86)

**Comment by dustymc** [comment]
Oh boy. I'm about 97% sure the "Do it for them!" bits would put us on the wrong side of a TACC audit.

>(USER accounts can, but not OPERATOR) 

Yea, I ran this by Lam and that's basically our conclusion as well. I need to link operators to people. If I can't use Agents for that (eg because what I've proposed here can't fly, and I don't think I'm getting any guidance from https://github.com/ArctosDB/arctos/issues/7649) then I need something else, and a TACC account is the most-obvious option that's not going to get Arctos locked down. If I am to continue using Agents then I clearly need more guardrails than we have now, and maybe that's just not in line with what The Community wants.

**Comment by adhornsby** [comment]
> @adhornsby can you please add a unique email account to https://arctos.database.museum/agent/21331627? And update their operator email as well?

Aw, my puppet account under my husband's name... but I get the problem.  Email corrected in both places!

**Comment by Jegelewicz** [comment]
> That is just one person's email (Ralph Imondi, agent id 21349246). It should be removed from all of the other agents it is associated with.

Done

**Comment by Jegelewicz** [comment]
@ewommack which one of these spellings is correct?

![image](https://github.com/ArctosDB/arctos/assets/5725767/44c8ce63-f4fd-4647-aa80-30787c5ecbff)

**Comment by Jegelewicz** [comment]
@mvzhuang can you add appropriate email addresses to these agents?

Bethany   Ramos
Lindsay Schodek

**Comment by mvzhuang** [comment]
Done!

**Comment by Jegelewicz** [comment]
@AdrienneRaniszewski can you add unique emails to the following agents?

Ary C.   Perez
Elisabetta Bastai
Elizabeth A. Racehorse
MSB DOM Volunteer
Serina A. Altamirano

**Comment by dustymc** [comment]
>This is definitely a problem

But not all of the problem, https://arctos.database.museum/agent/21339909?deets=true also exists.

**Comment by Jegelewicz** [comment]
@dustymc I cleaned up a bunch of these. What remains are things I cannot do anything about because it seems that the email address is generic to cover two people OR there are operators involved and I don't want to mess those up. I've tasked some people in this issue with some of the remaining things, but some of them I am not sure who to check in with.

**Comment by Jegelewicz** [comment]
Removed advice to create accounts for users from documentation.

**Comment by Jegelewicz** [comment]
@wellerjes please add unique email addresses to these agents

Ariel   Wagner
Daphne Onsay
Emmett Glazbrook
Olivia Cimino

**Comment by Jegelewicz** [comment]
@AJLinn @DellaCHall did one of these people leave the position at Museum of the Aleutians?

![image](https://github.com/ArctosDB/arctos/assets/5725767/023b039e-6c62-40a9-94e8-957923beaa60)

**Comment by AdrienneRaniszewski** [comment]
I don’t have emails for those folks, but I removed my address. MSB DOM Volunteer hasn’t been used in years; it can be deleted.


~ Adrienne


Collections Associate | Staff Manager
MSB Division of Mammals
505•277•5369




From: Teresa Mayfield-Meyer ***@***.***>
Date: Wednesday, May 22, 2024 at 2:14 PM
To: ArctosDB/arctos ***@***.***>
Cc: Adrienne Raniszewski ***@***.***>, Mention ***@***.***>
Subject: Re: [ArctosDB/arctos] Feature Request - make email unique (Issue #7803)
  [EXTERNAL]

@AdrienneRaniszewski<https://github.com/AdrienneRaniszewski> can you add unique emails to the following agents?

Ary C. Perez
Elisabetta Bastai
Elizabeth A. Racehorse
MSB DOM Volunteer
Serina A. Altamirano

**Comment by ewommack** [comment]
> which one of these spellings is correct?

Urrrghh my bad. Bad hand-writing on Accession forms created two different people, apologies. I double checked the name in the UWY email list, and marked the Cuthbent-Millett as the bad duplicate of, and then went and corrected the Accession the bad name was attached to.

**Comment by dustymc** [comment]
>cannot do anything about

Fresh data in the internal issue, in case it's handy.

I can do whatever, just not sure if I should. I really need some big-picture guidance. Can I reasonably expect to use agents as a way to link people to operators (in which case I'll do whatever's necessary to get us to and keep us in a workable state), or do I need to be looking at something like TACC accounts for that? Moving to TACC for auth would let them deal with all of this stuff, and allow us to access things like 2FA. (Which I'm sure everyone would hate, but it's likely still a significantly more secure mechanism.)

**Comment by AJLinn** [comment]
> @AJLinn @DellaCHall did one of these people leave the position at Museum of the Aleutians?

Karen was the contact in 2022:
![image](https://github.com/ArctosDB/arctos/assets/17605945/23aff8d3-ba27-4372-b072-5717b5c0aeb8)

The Museum of the Aleutians, as in many small community museums, have email addresses for staff positions, rather than the individual who occupies them (not unlike Arctos’ officer position emails). As a result, this unique email requirement for agents (like with the “family” emails some of our older donors still have) will create a problem in this current proposal.

I’ll have to go thru my emails in more detail once I”m in the office to get the start/end dates of previous MOA contacts.

**Comment by wellerjes** [comment]
> @wellerjes please add unique email addresses to these agents
> 
> Ariel Wagner Daphne Onsay Emmett Glazbrook Olivia Cimino

Deleted collectionsinfo from their agent records.

**Comment by dustymc** [comment]
Thanks everyone!

I cleaned up the final few, email is now required to be unique.

Note that some active operators are without email, those accounts will auto-lock soon-ish.

```
   username    |     last_login      |         rolvaliduntil         
---------------+---------------------+-------------------------------
 mvacanti      | 2024-05-28 00:00:00 | 2024-11-24 14:14:01.749448-06
 ocimino       | 2024-05-22 00:00:00 | 2024-11-18 08:18:37.009041-06
 tholcomb      | 2024-05-09 00:00:00 | 2024-08-21 13:17:48.764293-05
 cjohnson_dmns | 2024-05-28 00:00:00 | 2024-11-24 08:52:30.000566-06
 eglazbrook    | 2024-05-02 00:00:00 | 2024-10-29 12:15:51.449883-05
```
